chipseq_Intro
=======================

```
processing file: chipseq_Intro.Rpres
Loading required package: ggplot2
Loading required package: DiffBind
Loading required package: GenomicRanges
Loading required package: BiocGenerics
Loading required package: parallel

Attaching package: 'BiocGenerics'

The following objects are masked from 'package:parallel':

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB

The following objects are masked from 'package:stats':

    IQR, mad, xtabs

The following objects are masked from 'package:base':

    anyDuplicated, append, as.data.frame, as.vector, cbind,
    colnames, do.call, duplicated, eval, evalq, Filter, Find, get,
    grep, grepl, intersect, is.unsorted, lapply, lengths, Map,
    mapply, match, mget, order, paste, pmax, pmax.int, pmin,
    pmin.int, Position, rank, rbind, Reduce, rownames, sapply,
    setdiff, sort, table, tapply, union, unique, unlist, unsplit

Loading required package: S4Vectors
Loading required package: stats4
Loading required package: IRanges
Loading required package: GenomeInfoDb
Loading required package: SummarizedExperiment
Loading required package: Biobase
Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.

Loading required package: limma

Attaching package: 'limma'

The following object is masked from 'package:BiocGenerics':

    plotMA

Loading required package: GenomicAlignments
Loading required package: Biostrings
Loading required package: XVector
Loading required package: Rsamtools
Loading required package: locfit
locfit 1.5-9.1 	 2013-03-22

Attaching package: 'locfit'

The following objects are masked from 'package:GenomicAlignments':

    left, right

Loading required package: DBI

Using Sample as id variables
Using Sample as id variables
Scale for 'x' is already present. Adding another scale for 'x', which will replace the existing scale.
Using Sample as id variables
Scale for 'x' is already present. Adding another scale for 'x', which will replace the existing scale.
Loading required package: Rcpp
Loading required package: RcppArmadillo
Loading required package: grid
Loading required package: VennDiagram
Loading required package: futile.logger
Loading required package: GenomicFeatures
Loading required package: AnnotationDbi
Loading required package: org.Mm.eg.db

'select()' returned many:many mapping between keys and columns
'select()' returned many:many mapping between keys and columns
'select()' returned many:many mapping between keys and columns

KEGG.db contains mappings based on older data because the original
  resource was removed from the the public domain before the most
  recent update was produced. This package should now be
  considered deprecated and future versions of Bioconductor may
  not have it available.  Users who want more current data are
  encouraged to look at the KEGGREST or reactome.db packages

Loading required package: BiasedUrn
Loading required package: geneLenDataBase
Loading mm9 length data...
Fetching GO annotations...
For 15311 genes, we could not find any categories. These genes will be excluded.
To force their use, please run with use_genes_without_cat=TRUE (see documentation).
This was the default behavior for version 1.15.1 and earlier.
Calculating the p-values...
Loading mm9 length data...
Fetching GO annotations...
```
