---
title: "ChIP-seq Identifying Motifs - practical"
author: "MRC Clinical Sciences Bioinformatics"
output: html_document
---

In this practical we will read in two replicate sets of peaks from the Myc Encode dataset and extract sequences underneath subsets of peaks. We will write these sequences out to a FASTA file for us with Meme-ChIP.

* First lets read in the two Myc Mel peakset replicates and create the common peakset as we did for our previous exercise


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T,eval = T)
```

```{r,collapse=T,echo=F}
# Load the GenomicRanges Library .. here is suppress messages for a cleaner document
suppressPackageStartupMessages(
  library(GenomicRanges)
)
suppressPackageStartupMessages(
  library(BSgenome)
)
suppressPackageStartupMessages(
  library(BSgenome.Mmusculus.UCSC.mm9)
)

```
```{r,collapse=T}
  library(GenomicRanges)
  library(BSgenome)
  library(BSgenome.Mmusculus.UCSC.mm9)

```

```{r,collapse=T}

melPeak_Rep1 <- read.delim("data/MacsPeaks/mycmelrep1_peaks.xls",sep="\t",comment.char = "#")
melPeak_Rep2 <- read.delim("data/MacsPeaks/mycmelrep2_peaks.xls",sep="\t",comment.char = "#")

melRep1_GR <- GRanges(
                  seqnames=melPeak_Rep1[,"chr"],
                  IRanges(melPeak_Rep1[,"start"],
                  melPeak_Rep1[,"end"]
                  )
                )

mcols(melRep1_GR) <- melPeak_Rep1[,c("abs_summit", "fold_enrichment")]


melRep2_GR <- GRanges(
                  seqnames=melPeak_Rep2[,"chr"],
                  IRanges(melPeak_Rep2[,"start"],
                  melPeak_Rep2[,"end"]
                  )
                )

mcols(melRep2_GR) <- melPeak_Rep2[,c("abs_summit", "fold_enrichment")]

melRep1_GR_InRep1AndRep2 <- melRep1_GR[melRep1_GR %over% melRep2_GR]
melRep1_GR_InRep1AndRep2
```
* Now we can rank them by their fold enrichment, select the top 500 peaks and resize these peaks to 200bp around centre.

```{r,collapse=T}
melRep1_GR_InRep1AndRep2 <- melRep1_GR_InRep1AndRep2[ order(melRep1_GR_InRep1AndRep2$fold_enrichment,decreasing=T)
                         ]
top500_melRep1_Rep1AndRep2 <- melRep1_GR_InRep1AndRep2[1:500,]
top500_melRep1_1And2_Resized <- resize(top500_melRep1_Rep1AndRep2,200,fix = "center")
top500_melRep1_1And2_Resized[1:3,]
```
* Extract the sequences underneath the file and write them to FASTA file in you working directory. Inspect the file in notepad.

```{r,collapse=T}
genome <- BSgenome.Mmusculus.UCSC.mm9

seqlevelsStyle(top500_melRep1_1And2_Resized) <- "UCSC"

top500_melRep1_1And2_Resized_Seq <- getSeq(genome,top500_melRep1_1And2_Resized)
names(top500_melRep1_1And2_Resized_Seq) <- paste0("peak_",seqnames(top500_melRep1_1And2_Resized),"_",
                                         start(top500_melRep1_1And2_Resized),
                                         "-",
                                         end(top500_melRep1_1And2_Resized))

top500_melRep1_1And2_Resized_Seq[1:4,]

writeXStringSet(top500_melRep1_1And2_Resized_Seq,file="top500_melRep1_1And2.fa")
```

* Upload the sequences to Meme-ChIP and review the results when complete.

**Once you have uploaded the data you can preview the results [here](http://mrccsc.github.io/appMEMECHIP_4.11.11457523923197-718963456/meme-chip.html)**

** Bonus question - **Recentre top 500 peaks around their absolute summit**, resize to 200, extract sequences and submit to Meme-ChIP. Compare results between peaksets.

**Once you have uploaded the data you can preview the results [here](http://mrccsc.github.io/appMEMECHIP_4.11.114575239877311434368117/meme-chip.html)**
```{r,collapse=T}

start(top500_melRep1_Rep1AndRep2) <- end(top500_melRep1_Rep1AndRep2) <- top500_melRep1_Rep1AndRep2$abs_summit
top500_melRep1_1And2_Resized <- resize(top500_melRep1_Rep1AndRep2,200,fix = "center")


genome <- BSgenome.Mmusculus.UCSC.mm9

seqlevelsStyle(top500_melRep1_1And2_Resized) <- "UCSC"

top500_melRep1_1And2_Resized_Seq <- getSeq(genome,top500_melRep1_1And2_Resized)
names(top500_melRep1_1And2_Resized_Seq) <- paste0("peak_",seqnames(top500_melRep1_1And2_Resized),"_",
                                         start(top500_melRep1_1And2_Resized),
                                         "-",
                                         end(top500_melRep1_1And2_Resized))

writeXStringSet(top500_melRep1_1And2_Resized_Seq,file="abssummit_top500_melRep1_1And2.fa")
```

