---
title: "ChIP-seq Working With Peaks - practical"
author: "MRC CSC Bioinformatics Core"
date: "22 February 2016"
output: html_document
---

In this practical we will read in two replicate sets of peaks from the Myc Encode dataset,filter by fold enrichment over input and find the number of peaks common to both replicates and those unique to either.


* First read in each of the Mel replicate peak sets found in MacsPeaks directory and create a GRanges object containing peak locations for each.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T,eval = T)
```

```{r,collapse=T}
# Load the GenomicRanges Library .. here is suppress messages for a cleaner document
suppressPackageStartupMessages(
  library(GenomicRanges)
  )

melPeak_Rep1 <- read.delim("data/MacsPeaks/mycmelrep1_peaks.xls",sep="\t",comment.char = "#")
melPeak_Rep2 <- read.delim("data/MacsPeaks/mycmelrep2_peaks.xls",sep="\t",comment.char = "#")

melRep1_GR <- GRanges(
                  seqnames=melPeak_Rep1[,"chr"],
                  IRanges(melPeak_Rep1[,"start"],
                  melPeak_Rep1[,"end"]
                  )
                )

mcols(melRep1_GR) <- melPeak_Rep1[,c("abs_summit", "fold_enrichment")]

melRep1_GR

melRep2_GR <- GRanges(
                  seqnames=melPeak_Rep2[,"chr"],
                  IRanges(melPeak_Rep2[,"start"],
                  melPeak_Rep2[,"end"]
                  )
                )

mcols(melRep2_GR) <- melPeak_Rep2[,c("abs_summit", "fold_enrichment")]

melRep2_GR

```

Now you have your peak sets as GRanges, you can start to work with the peaks in R.

* Find the number of peaks in each replicate.

* Find the number of peaks in each replicate on chromosome 4.

* Find the number of peaks on chromosome 4 which have a 5 fold enrichment above input.

```{r, warnings=F,collapse=T}
# Number of peaks

length(melRep1_GR)

length(melRep2_GR)

# Number of peaks on chromosome 4

# Using table on logical vector

table(seqnames(melRep1_GR) %in% "4")

table(seqnames(melRep2_GR) %in% "4")

# Indexing and recounting

length(melRep1_GR[seqnames(melRep1_GR) %in% "4"])

length(melRep2_GR[seqnames(melRep2_GR) %in% "4"])


# Number of peaks on chromosome 4 and with 5 fold enrichment above input

# Using table on logical vector

table(seqnames(melRep1_GR) %in% "4" & melRep1_GR$fold_enrichment > 5)

table(seqnames(melRep2_GR) %in% "4" & melRep2_GR$fold_enrichment > 5)

# Indexing and recounting

length(melRep1_GR[seqnames(melRep1_GR) %in% "4" & melRep1_GR$fold_enrichment > 5])

length(melRep2_GR[seqnames(melRep2_GR) %in% "4" & melRep2_GR$fold_enrichment > 5])

```


Now we can use our GRanges to identify the peak in common or unique to replicates.

* Find the number of replicate 1 peaks common between replicates.
* Find the number of unique peaks in replicate 1

```{r, warnings=F,collapse=T}

# Using table

table(melRep1_GR %over% melRep2_GR)

# Using index and recounting
length(melRep1_GR[melRep1_GR %over% melRep2_GR])


# Using table

table(!melRep1_GR %over% melRep2_GR)

# Using index and recounting
length(melRep1_GR[!melRep1_GR %over% melRep2_GR])

commonMelPeaks <- melRep1_GR[melRep1_GR %over% melRep2_GR]
```

For a bonus question. 

* Create a new GRanges of the summit locations from Mel Myc Peaks replicate 1.

```{r, warnings=F,collapse=T}

# Using table

melRep1_GRSummits <- melRep1_GR

start(melRep1_GRSummits) <- end(melRep1_GRSummits) <- melRep1_GR$abs_summit

melRep1_GRSummits
```

